pkg/dockerregistry/server/app.go

```diff
@@ -0,0 +1,143 @@
+	app := &App{
+		ctx:            ctx,
+		registryClient: registryClient,
+		extraConfig:    extraConfig,
+		writeLimiter:   writeLimiter,
+	}
+
+	cache, err := newDigestToRepositoryCache(defaultDigestToRepositoryCacheSize)
+	if err != nil {
+		panic(err)
+	}
+	app.cachedLayers = cache
+
+	weaveAppIntoConfig(app, dockerConfig)
+
+	for _, middleware := range dockerConfig.Middleware["repository"] {
+		if middleware.Name == middlewareOpenShift {
+			rc, err := newRepositoryConfig(ctx, middleware.Options)
+			if err != nil {
+				context.GetLogger(ctx).Fatalf("error configuring the repository middleware: %s", err)
+			}
+			app.repositoryConfig = rc
+			app.quotaEnforcing = newQuotaEnforcingConfig(ctx, os.Getenv(EnforceQuotaEnvVar), os.Getenv(ProjectCacheTTLEnvVar), middleware.Options)
+			break
+		}
+	}
+
+	dockerApp := handlers.NewApp(ctx, dockerConfig)
+
+	// Add a token handling endpoint
+	if dockerConfig.Auth.Type() == middlewareOpenShift {
+		tokenRealm, err := TokenRealm(dockerConfig.Auth[middlewareOpenShift])
+		if err != nil {
+			context.GetLogger(dockerApp).Fatalf("error setting up token auth: %s", err)
+		}
+		err = dockerApp.NewRoute().Methods("GET").PathPrefix(tokenRealm.Path).Handler(NewTokenHandler(ctx, registryClient)).GetError()
+		if err != nil {
+			context.GetLogger(dockerApp).Fatalf("error setting up token endpoint at %q: %v", tokenRealm.Path, err)
+		}
+		context.GetLogger(dockerApp).Debugf("configured token endpoint at %q", tokenRealm.String())
+	}
+
+	app.registerBlobHandler(dockerApp)
+
+	// Registry extensions endpoint provides extra functionality to handle the image
+	// signatures.
+	RegisterSignatureHandler(dockerApp)
+
+	// Registry extensions endpoint provides prometheus metrics.
+	if extraConfig.Metrics.Enabled {
+		if len(extraConfig.Metrics.Secret) == 0 {
+			context.GetLogger(dockerApp).Fatalf("openshift.metrics.secret field cannot be empty when metrics are enabled")
+		}
+		RegisterMetricHandler(dockerApp)
+	}
+
+	// Advertise features supported by OpenShift
+	if dockerApp.Config.HTTP.Headers == nil {
+		dockerApp.Config.HTTP.Headers = http.Header{}
+	}
+	dockerApp.Config.HTTP.Headers.Set("X-Registry-Supports-Signatures", "1")
+
+	dockerApp.RegisterHealthChecks()
+
+	return dockerApp
+}
+
+func (app *App) newRepository(ctx context.Context, repo distribution.Repository, options map[string]interface{}) (distribution.Repository, error) {
+	if app.driver == nil {
+		return nil, fmt.Errorf("configuration error: the storage driver middleware %q is not activated", middlewareOpenShift)
+	}
+
+	if app.registry == nil {
+		return nil, fmt.Errorf("configuration error: the registry middleware %q is not activated", middlewareOpenShift)
+	}
+
+	registryOSClient, err := app.registryClient.Client()
+	if err != nil {
+		return nil, err
+	}
+
+	return app.newRepositoryWithClient(ctx, registryOSClient, repo)
```

Comment:
Why do we need second method that just accepts client obtained by the first, which could be obtained by the second as well?